gax/retry_loop.h

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gax/backoff_policy.h"
+#include "gax/call_context.h"
+#include "gax/retry_policy.h"
+#include "gax/status.h"
+#include <chrono>
+#include <functional>
+#include <thread>
+#include <type_traits>
+
+#ifndef GAPIC_GENERATOR_CPP_GAX_RETRY_LOOP_H_
+#define GAPIC_GENERATOR_CPP_GAX_RETRY_LOOP_H_
+
+namespace google {
+namespace gax {
+
+template <typename RequestT, typename ResponseT>
+gax::Status MakeRetryCall(
+    CallContext& context, RequestT const& request, ResponseT* response,
+    std::function<Status(CallContext&, RequestT const&, ResponseT*)> next_stub,
```

Comment:
Since this is a template already, why not make `next_stub` a functor (as in `F&& next_stub`) anyway?  That can save you allocations and inline the function in the critical path.

Reply:
I was on the fence about that. The functor saves an allocation and lets the compiler inline the function, but type checking becomes more difficult. I'm sure there's a way of handling that with `std::enable_if` and type traits, but I couldn't figure out how.

I'll try moving it into a template parameter and seeing how bad the compilation errors get.

Reply:
>  I'm sure there's a way of handling that with `std::enable_if` and type traits, but I couldn't figure out how.

In C++17:

```
std::enable_if<std::is_invocable_v<FunctorType, CallContext& RequestT const&, ResponseT*>, int>
```

In C++11:

```
std::enable_if<typename google::cloud::internal::is_invocable<FunctorType, CallContext& RequestT const&, ResponseT*>::value, int>
```

Reply:
+1 for a functor (the lambdas seem just proxying the call to the function, but are very verbose and less efficient).

But what is `google::cloud::internal::is_invocable`? Is it from google-cloud-cpp? If yes, are we going to copy it to cpp gax?